packages/ic-signature-verification/src/lib.rs

```diff
@@ -0,0 +1,163 @@
+use candid::Principal;
+use ic_cbor::CertificateToCbor;
+use ic_certification::{
+    leaf, Certificate, Delegation, HashTree, LookupResult, SubtreeLookupResult,
+};
+use ic_verify_bls_signature::verify_bls_signature;
+use serde::{Deserialize, Serialize};
+use serde_bytes::ByteBuf;
+use sha2::{Digest, Sha256};
+
+pub const IC_STATE_ROOT_DOMAIN_SEPARATOR: &[u8; 14] = b"\x0Dic-state-root";
+
+use ic_canister_sig_creation::{extract_raw_root_pk_from_der, CanisterSigPublicKey};
+
+/// Verifies that `signature` is a valid canister signature on `message`.
+/// https://internetcomputer.org/docs/current/references/ic-interface-spec#canister-signatures
+pub fn verify_canister_sig(
+    message: &[u8],
+    signature_cbor: &[u8],
+    public_key_der: &[u8],
+    ic_root_public_key_raw: &[u8],
+) -> Result<(), String> {
+    let signature = parse_signature_cbor(signature_cbor)?;
+    let public_key = CanisterSigPublicKey::try_from(public_key_der)
+        .map_err(|e| format!("failed to parse canister sig public key: {}", e))?;
+    let certificate = check_certified_data_and_get_certificate(&signature, &public_key, message)?;
+    verify_certificate(&certificate, public_key.canister_id, ic_root_public_key_raw)
+}
+
+fn check_certified_data_and_get_certificate(
+    signature: &CanisterSignature,
+    canister_sig_pk: &CanisterSigPublicKey,
+    msg: &[u8],
+) -> Result<Certificate, String> {
+    let certificate = Certificate::from_cbor(&signature.certificate)
```

Comment:
I mainly find it odd that one library (`ic-certification`) implements the types, and another library (`ic-cbor`) implements the parsing. My impression is it would be best to have this all handled transparently directly in `ic-certification` (which may use `ic-cbor` underneath if need be). Maybe @nathanosdev has some insights here.

I'm OK with keeping things as if for now.

Reply:
The main purpose for `ic-cbor` was to provide a deserialization that would keep a smaller bundle size at the end when compiling to WASM and binding to JavaScript. `serde_cbor` added quite a significant increase to file size.

This was important for the Service Worker, which we wanted to keep as small as possible, but since the Service Worker is gone now, we don't have that requirement anymore. I'd be in favor of not using it and eventually removing it in favor of `serde_cbor`.

Reply:
Done.